lib/SyTest/Identity/Server.pm

```diff
@@ -0,0 +1,121 @@
+
+   $self->{keys} = {
+      "ed25519:0" => encode_base64_unpadded( $self->{public_key} ),
+   };
+
+}
+
+sub on_request
+{
+   my $self = shift;
+   my ( $req ) = @_;
+
+   my $path = $req->path;
+   my %resp;
+
+   if( $path eq "/_matrix/identity/api/v1/pubkey/isvalid" ) {
+      my $user_agent = $req->header( "User-Agent" );
+      if( defined $self->{isvalid_needs_useragent} and !$user_agent =~ m/\Q$self->{isvalid_needs_useragent}/ ) {
+         die "Wrong useragent made /isvalid request";
+      }
+      $resp{valid} = (any { $_ eq $req->query_param("public_key") } values $self->{keys}) ?
+         JSON::true : JSON::false;
+      $req->respond_json( \%resp );
+   } elsif( my ( $key_name ) = $path =~ m#^/_matrix/identity/api/v1/pubkey/([^/]*)$# ) {
+      if( defined $self->{keys}->{$key_name} ) {
+         $resp{public_key} = $self->{keys}{$key_name};
+      }
+      $req->respond_json( \%resp );
+   } elsif( $path eq "/_matrix/identity/api/v1/lookup" ) {
+      my ( $req ) = @_;
+      my $medium = $req->query_param( "medium" );
+      my $address = $req->query_param( "address" );
+      if ( !defined $medium || !defined $address ) {
+         $req->respond( HTTP::Response->new( 400, "Bad Request", [ Content_Length => 0 ] ) );
+         return
+      }
+      my $mxid = $self->{bindings}{$address};
+      if ( "email" eq $medium && defined $mxid ) {
+         $resp{medium} = $medium;
+         $resp{address} = $address;
+         $resp{mxid} = $mxid;
+
+         my $sock = $self->read_handle;
+         my $name = sprintf "%s:%d", $sock->sockhostname, $sock->sockport;
+         sign_json( \%resp,
+            secret_key => $self->{private_key},
+            origin => $name,
+            key_id => "ed25519:0",
+         );
+      }
+      $req->respond_json( \%resp );
+   } elsif( $path eq "/_matrix/identity/api/v1/store-invite" ) {
+      my $body = $req->body_from_form;
+      my $medium = $body->{medium};
+      my $address = $body->{address};
+      my $sender = $body->{sender};
+      my $room_id = $body->{room_id};
+      if( !( defined $body->{medium} and defined $address and defined $sender and defined $room_id ) ) {
+         $req->respond( HTTP::Response->new( 400, "Bad Request", [ Content_Length => 0 ] ) );
+         return
+      }
+      my $token = $self->{expected_tokens}{$medium.$address.$sender.$room_id};
+      if( !defined $token ) {
+         $req->respond( HTTP::Response->new( 500, "Internal Server Error", [ Content_Length => 0 ] ) );
+         return
+      }
+      $resp{token} = $token;
+      $resp{public_key} = $self->{keys}{"ed25519:0"};
+      $req->respond_json( \%resp );
+   } else {
+      $req->respond( HTTP::Response->new( 404, "Not Found", [ Content_Length => 0 ] ) );
+   }
+}
+
+sub stub_token {
+   my $self = shift;
+   my ( $token, $medium, $invitee_email, $inviter_mxid, $room_id ) = @_;
+
+   $self->{expected_tokens}{$medium.$invitee_email.$inviter_mxid.$room_id} = $token;
```

Comment:
It probably doesn't matter in a test like this, but a more robust mechanism for building these keys that's not so fragile as simple string concatenation is to `join` on some character not allowable in the strings themselves - `\0` is often a likely candidate:

    $self->{expected_tokens}{ join "\0", $medium, $invitee_email, $inviter_mxid, $room_id } = $token;